tests/requestparsing/basicauth_test.go

```diff
@@ -0,0 +1,170 @@
+				"Authorization: xasic UGVsbGU6UGFzc3dvcmQ6UGFzc3dvcmQ=\r\n" +
+				"\r\n"),
+			want: testWant{
+				// Same Base64 as above.
+				headers:  map[string][]string{"Authorization": []string{"xasic UGVsbGU6UGFzc3dvcmQ6UGFzc3dvcmQ="}},
+				ok:       false,
+				username: "",
+				password: "",
+			},
+		},
+		{
+			name: "Ordering",
+			request: []byte("GET / HTTP/1.1\r\n" +
+				"Host: localhost:8080\r\n" +
+				// Base64 encoding of "AAA:aaa".
+				"Authorization: basic QUFBOmFhYQ==\r\n" +
+				// Base64 encoding of "BBB:bbb".
+				"Authorization: basic QkJCOmJiYg==\r\n" +
+				"\r\n"),
+			want: testWant{
+				// Base64 encoding of "AAA:aaa" and then of "BBB:bbb" in that order.
+				headers:  map[string][]string{"Authorization": []string{"basic QUFBOmFhYQ==", "basic QkJCOmJiYg=="}},
+				ok:       true,
+				username: "AAA",
+				password: "aaa",
+			},
+		},
+		{
+			name: "CasingOrdering1",
+			request: []byte("GET / HTTP/1.1\r\n" +
+				"Host: localhost:8080\r\n" +
+				// Base64 encoding of "AAA:aaa".
+				"Authorization: basic QUFBOmFhYQ==\r\n" +
+				// Base64 encoding of "BBB:bbb".
+				"authorization: basic QkJCOmJiYg==\r\n" +
+				"\r\n"),
+			want: testWant{
+				// Base64 encoding of "AAA:aaa" and then of "BBB:bbb" in that order.
+				headers:  map[string][]string{"Authorization": []string{"basic QUFBOmFhYQ==", "basic QkJCOmJiYg=="}},
+				ok:       true,
+				username: "AAA",
+				password: "aaa",
+			},
+		},
+		{
+			name: "CasingOrdering2",
+			request: []byte("GET / HTTP/1.1\r\n" +
+				"Host: localhost:8080\r\n" +
+				// Base64 encoding of "AAA:aaa".
+				"authorization: basic QUFBOmFhYQ==\r\n" +
+				// Base64 encoding of "BBB:bbb".
+				"Authorization: basic QkJCOmJiYg==\r\n" +
+				"\r\n"),
+			want: testWant{
+				// Base64 encoding of "AAA:aaa" and then of "BBB:bbb" in that order.
+				headers:  map[string][]string{"Authorization": []string{"basic QUFBOmFhYQ==", "basic QkJCOmJiYg=="}},
+				ok:       true,
+				username: "AAA",
+				password: "aaa",
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			resp, err := requesttesting.MakeRequest(context.Background(), tt.request, func(r *http.Request) {
+				if diff := cmp.Diff(tt.want.headers, map[string][]string(r.Header)); diff != "" {
+					t.Errorf("r.Header mismatch (-want +got):\n%s", diff)
+				}
+
+				username, password, ok := r.BasicAuth()
+				if ok != tt.want.ok {
+					t.Errorf("_, _, ok := r.BasicAuth() got: %v want: %v", ok, tt.want.ok)
+				}
+
+				if username != tt.want.username {
+					t.Errorf("username, _, _ := r.BasicAuth() got: %q want: %q", username, tt.want.username)
+				}
+
+				if password != tt.want.password {
+					t.Errorf("_, password, _ := r.BasicAuth() got: %q want: %q", password, tt.want.password)
+				}
+			})
+			if err != nil {
+				t.Fatalf("MakeRequest() got err: %v", err)
+			}
+
+			if !bytes.HasPrefix(resp, []byte(statusOK)) {
+				got := string(resp[:bytes.IndexByte(resp, '\n')+1])
```

Comment:
Won't this give you the trailing newline as well? That will make the error message look weird, with a linebreak in between.

Reply:
I'm printing the error messages with `%q` so it will be escaped like this: `"HTTP/1.1 200 OK\r\n"`.
Should I removed the line breaks and the switch to `%s`?

Reply:
Ah right, %q escapes the newline characters.

I proposed in another comment to have a helper that extracts just the first line, without the \r\n.